### PR TITLE
no-semi comments

### LINKED ISSUE #1245: Semicolon is inserted in front of a comment, instead of the expression
[[online demo](https://prettier.github.io/prettier/#%7B%22content%22%3A%22let%20error%20%3D%20new%20Error(response.statusText)%3B%5Cn%2F%2F%20a%20comment%5Cn(error%3A%20any).response%20%3D%20response%5Cn%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Afalse%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%7D%7D)]

```js
let error = new Error(response.statusText);
// a comment
(error: any).response = response

// becomes

let error = new Error(response.statusText)
;// a comment
(error: any).response = response
```

I would expect the semicolon to be inserted at the beginning of the `(error: any)`, rather than in front of the comment.

### PR DESCRIPTION
**Pull request by vjeux**
Fixes #1245

### CODE REVIEW

**Comment by vjeux** [comment]
Okay, it doesn't work because needsParen is mutated during the print operation, which has been moved afterwards...

File: tests/comments/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -1112,7 +1112,7 @@ switch (node && node.type) {
   case "MethodDefinition":
     prop = node.key;
     break;
-
+
```
Comment (vjeux) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: This will be fixed by #1259

File: tests/no-semi/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -1,5 +1,86 @@
...
+
+/* comment */ (error: any).response = response
+
+x;
+
+(error: any).response = response; /* comment */
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+let error = new Error(response.statusText)
+// comment
+;(error: any).response = response
```
Comment (vjeux) [inline tests/no-semi/__snapshots__/jsfmt.spec.js.snap]: yay!

**Comment by vjeux** [comment]
No, it's not going to work but I can't find any real-world use case where it would be an issue. Can you think of one?

I can make up examples though: https://prettier.github.io/prettier/#%7B%22content%22%3A%22%3B(%5Cn%20%20%2F%2F%20a%5Cn%20%201%20-%201)%20*%2010%5Cn%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22all%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Afalse%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%7D%7D

**Comment by rattrayalex** [comment]
Thanks, that's helpful. I don't think it's high priority but I'll think
about it...

**Pull request merged**